clients/src/main/java/org/apache/kafka/common/requests/JoinGroupRequest.java

```diff
@@ -62,51 +60,103 @@ public ByteBuffer metadata() {
         }
     }
 
-    // v0 constructor
-    @Deprecated
-    public JoinGroupRequest(String groupId,
-                            int sessionTimeout,
-                            String memberId,
-                            String protocolType,
-                            List<ProtocolMetadata> groupProtocols) {
-        this(0, groupId, sessionTimeout, sessionTimeout, memberId, protocolType, groupProtocols);
-    }
+    public static class Builder extends AbstractRequest.Builder<JoinGroupRequest> {
+        private String groupId = null;
+        private int sessionTimeout = 0;
+        private int rebalanceTimeout = 0;
+        private String memberId = null;
+        private String protocolType = null;
+        private List<ProtocolMetadata> groupProtocols = null;
+
+        public Builder(String groupId, int sessionTimeout, String memberId,
```

Comment:
It would be nice to be consistent about whether a field should be included in the constructor or whether it should have a setter. I don't know that we have any cases that call for mutating a field that was initialized in the constructor, so I would prefer to just leave off the setters in that case and make the field final.